accepted/2024/[TS]PMUI-Readme-rendering.md

```diff
@@ -0,0 +1,213 @@
+
+### Technical explanation
+#### Rendering Markdown
+
+We will use the [Microsoft.VisualStudio.Markdown.Platform](https://dev.azure.com/azure-public/vside/_artifacts/feed/vs-impl/NuGet/Microsoft.VisualStudio.Markdown.Platform/overview/18.0.39-preview-g659b28ccd8) package to render the README in the IDE. This will allow us to leverage a centralized tool for rendering markdown in the IDE. 
+
+A new instance of the [IMarkdownPreview](https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=%2Fsrc%2FProductivity%2FMarkdownLanguageService%2FImpl%2FMarkdown.Platform%2FPreview%2FIMarkdownPreview.cs) can be created using [PreviewBuilder](https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=/src/Productivity/MarkdownLanguageService/Impl/Markdown.Platform/Preview/PreviewBuilder.cs).
+
+We can use the Preview builder as follows:
+```C#
+//This creates a new instance of the preview builder
+var markdownPreview = new PreviewBuilder().Build();
+
+//We update the current markdown being rendered by calling "UpdateContentAsync"
+markdownPreview.UpdateContentAsync(markDown ?? string.Empty, ScrollHint.None)
+
+//IMarkdownPreview.VisualElement contains the FrameworkElement to be passed to the view
+MarkdownPreviewControl = markdownPreview.VisualElement
+```
+#### Locating the README
+
+Create a new implementation of the `INuGetResource` interface, `ReadmeUriTemplateResource`.
+This will only be available for sources which have implemented the new `ReadmeUriTemplate` resource type.
+
+Update `IPackageSearchMetadata` to include the `ReadmeFileUrl` field. 
+`LocalPackageSearchMetadata` will populate the field with the location of the README, if available, on the local disk.
+When a server implements the new `RegistrationsBaseUrl` resource type, then the `ReadmeFileUrl` field will be returned from the server if a README is available and be deserialized from the response.
+When a server implements the `ReadmeUriTemplate` resource type and `ReadmeFileUrl` is empty, it will be populated with the url to the README. 
+
+#### Downloading the README
+
+Update the `INuGetPackageFileService` to add a new method for downloading the README. 
+
+## Drawbacks
+
+MarkdownPreview control currently marked as obsolete since the interface has not been finalized.
+So when an upgrade is made we may have to change how we use the control.
+
+WebView2 controls always render ontop of other controls in the view.
+[Secnario 25254665](https://microsoft.visualstudio.com/Edge/_workitems/edit/25254665).
+PM UI needs to be updated to ensure items don't scroll off screen.
+
+`INuGetPackageFileService` in it's current implementation cannot download files from feeds that require authentication.
+
+## Rationale and alternatives
+By using an existing control we maintain consistency throughout the IDE and can rely on the owner to fix any bugs with the control.
```

Comment:
maintain consistency with what?

VS extension readmes might allow different markdown to what nuget.org does for package readmes, so it seems to me that consistency with nuget.org's readme rendering is more important than consistency with readmes of some unrelated component.

I feel like the advantage of using the existing component is quicker development, and hopefully less maintainance.

The drawbacks are inconsistencies if the markdown dialect is different to what nuget.org already supports. And if the team who own the component aren't open to changing the markdown dialect for nuget, then it reduces NuGet's ability to make improvements to the supported markdown dialect.

I feel like this description of "alternatives and alternates" isn't giving a fair or accurate pro-con list.  In fact, reading this spec, it's not clear to me if the drawbacks were even considered.

Reply:
I filled it out a bit. 

When I said consistency, I meant how the Markdown is rendered, a markdown heading should appear the same throughout the IDE regardless of which experience you're in. You're right about the dialect though, I added that bit. I don't think it's that big of a concern though, unless NuGet.org decides to move to a completely new dialect... which would be a bunch of work for us even if we owned the control outright.